### ISSUE TITLE
New lint: prefer bool::then

### ISSUE THREAD

**Issue by magurotuna** [rust-lang/rust-clippy#6760]
### What it does

Checks for if-else that could be rewritten to [`bool::then`][boolthen]

[boolthen]: https://doc.rust-lang.org/std/primitive.bool.html#method.then

### Categories (optional)

- Kind: ~~style~~ restriction

Reduces redundancy and lines of code by leveraging a new stable API [`bool::then`][boolthen]

### Drawbacks

None.

### Example

```rust
let a = if foo {
    println!("hello foo!");
    Some(bar())
} else {
    None
};
```

Could be written as:

```rust
let a = foo.then(|| {
    println!("hello foo!");
    bar()
});
```

Note that the following cannot be rewritten using [`bool::then`](boolthen) because it has some action in `else` block:

```rust
let a = if foo {
    println!("hello foo!");
    Some(bar())
} else {
    println!("else!");
    None
};
```

**Comment by magurotuna** [comment]
I would like to implement it myself. If there's any concern or negative opinion on this lint, please let me know.
@rustbot claim

**Comment by camsteffen** [comment]
We definitely should have this. But I have a feeling that `bool::then` is "a little too functional" for some people and so it should not warn-by-default. So either pedantic or restriction. I would lean towards restriction.

**Comment by magurotuna** [comment]
Sounds reasonable. I'm in favor of choosing "restriction" too.  I'll go with it.
